### DISCUSSION TITLE
RCP radio capabilities requirements

### DISCUSSION THREAD

**Discussion by jkcko** [openthread/openthread#9859] [Q&A]
I am implementing the radio PAL API to build RCP and NCP devices but the first time I started the RCP it basically said I have to implement the OT_RADIO_CAPS_TRANSMIT_SEC and OT_RADIO_CAPS_TRANSMIT_TIMING capabilities.

As far as enabling the transmit security I see the simulation and efr32 implementations call a few otMacFrame* API functions that appear to actually perform the encryption.  Does the RCP build actually include any MAC library/utility functions?  Would it be sufficient to call these functions in my own radio PAL.  Are there any other guidelines or considerations I should be looking out for?

Where can I find more details about the transmit timing capabilities requirements?

**Comment by abtink** [comment]
> Does the RCP build actually include any MAC library/utility functions?

Yes. It does.
The source files included in RCP build:
https://github.com/openthread/openthread/blob/33574ad4175ffb088bcca047f4c8d5fb240d1495/src/core/CMakeLists.txt#L262-L296

> Are there any other guidelines or considerations I should be looking out for?

The expected behavior of platform APIs is documented in the platform header (e.g. [`openthread/platform/radio.h`](https://github.com/openthread/openthread/blob/33574ad4175ffb088bcca047f4c8d5fb240d1495/include/openthread/platform/radio.h)).

Also highly suggest checking existing implementations for different radio chips (repos in https://github.com/openthread)

**Reply by jkcko** [reply]
I proceeded to implement transmit timing using the mTxInfo.mTxDelayBaseTime and mTxDelay but the fields appear to never be set to anything but 0.  Digging around the code it seems it may only be used if CSL is enabled.  Is this correct?

I've also dug around the ot-efr32, ot-cc2538, ot-rtos repos and the radio caps range from NONE to some combination of ACK_TIMEOUT | ENERGY_SCAN | TRANSMIT_RETRIES | CSMA_BACKOFF but none have TRANSMIT_TIMING.  Do none of those platforms support building the RCP target for use with a linux based border router?

**Reply by jwhui** [reply]
`TRANSMIT_TIMING` is used with CSL.

Checkout `ot-nrf528xx` for an example with `TRANSMIT_TIMING`:

https://github.com/openthread/ot-nrf528xx/blob/c306b4c4dad12907b440305283334925de1a33b9/src/src/radio.c#L597
